haystack/utils/base_serialization.py

```diff
@@ -172,69 +178,103 @@ def _convert_to_basic_types(value: Any) -> Any:
-        return cls(_convert_to_basic_types(v) for v in value)
+        return [_convert_to_basic_types(v) for v in value]
 
     # primitive
     return value
 
 
-# TODO: Make this function public once its implementation is finalized and tested
-def _deserialize_value_with_schema(serialized: Dict[str, Any]) -> Dict[str, Any]:
+def _deserialize_value_with_schema(serialized: Dict[str, Any]) -> Any:  # pylint: disable=too-many-return-statements, # noqa: PLR0911, PLR0912
     """
-    Deserializes a dictionary with schema information and data to original values.
+    Deserializes a value with schema information back to its original form.
 
     Takes a dict of the form:
       {
-         "schema": {
-            "numbers": {"type": "integer"},
-            "messages": {"type": "array", "items": {"type": "haystack.dataclasses.chat_message.ChatMessage"}},
-        },
-        "data": {
-            "numbers": 1,
-            "messages": [{"role": "user", "meta": {}, "name": None, "content": [{"text": "Hello, world!"}]}],
+         "serialization_schema": {"type": "integer"} or {"type": "object", "properties": {...}},
+         "serialized_data": <the actual data>
       }
 
     :param serialized: The serialized dict with schema and data.
-    :returns: The deserialized dict with original values.
+    :returns: The deserialized value in its original form.
     """
-    schema = serialized.get("serialization_schema", {})
-    data = serialized.get("serialized_data", {})
-
-    result: Dict[str, Any] = {}
-    for field, raw in data.items():
-        info = schema.get(field)
-        # no schema entry → just deep-deserialize whatever we have
-        if not info:
-            result[field] = _deserialize_value(raw)
-            continue
-
-        t = info["type"]
-
-        # ARRAY case
-        if t == "array":
-            item_type = info["items"]["type"]
-            reconstructed = []
-            for item in raw:
-                envelope = {"type": item_type, "data": item}
-                reconstructed.append(_deserialize_value(envelope))
-            result[field] = reconstructed
 
-        # PRIMITIVE case
-        elif t in ("null", "boolean", "integer", "number", "string"):
-            result[field] = raw
-
-        # GENERIC OBJECT
-        elif t == "object":
-            envelope = {"type": "object", "data": raw}
-            result[field] = _deserialize_value(envelope)
+    if not serialized or "serialization_schema" not in serialized or "serialized_data" not in serialized:
+        raise DeserializationError(
+            f"Invalid format of passed serialized payload. Expected a dictionary with keys "
+            f"'serialization_schema' and 'serialized_data'. Got: {serialized}"
+        )
+    schema = serialized["serialization_schema"]
+    data = serialized["serialized_data"]
+
+    schema_type = schema.get("type")
+
+    if not schema_type:
+        # for backward comaptability till Haystack 2.17 we use legacy implementation
```

Comment:
I think we can reduce this until `2.16`. Our deprecation policy is by default for only one version now. What do you think?